firebase-firestore/src/test/java/com/google/firebase/firestore/local/CountingQueryEngine.java

```diff
@@ -35,10 +40,17 @@
  * mutations read.
  */
 class CountingQueryEngine extends QueryEngine {
+  enum OverlayType {
+    Patch,
+    Set,
+    Delete
+  }
+
   private final QueryEngine queryEngine;
 
   private final int[] overlaysReadByCollection = new int[] {0};
   private final int[] overlaysReadByKey = new int[] {0};
+  private final Map<DocumentKey, OverlayType> overlayTypes = new HashMap();
```

Comment:
would it make more sense to store the overlay itself (`Map<DocumentKey, Mutation>`) and have the test check the overlay? If we do this, we don't need to define `OverlayType`, and we check not only the type but also the overlay content.

Reply:
It would if the mutation used as overlay is the same (or can be easily reconstructed manually). They often have different Preconditions (overlay mutation tend to have no preconditions at all), and does not work with ServerTimestamp for example.

Potentially we could try to construct overlays carefully in each call site, but IMO this is exposing too much implementation details in the tests. The contract here is applying Overlay gets the same documents as applying mutations, not Overlay == Mutations.

Reply:
fair enough.